src/base_obs.c

```diff
@@ -0,0 +1,329 @@
+
+/** SBP callback for when the base station sends us a message containing its
+ * known location.
+ * Stores the base station position in the global #base_pos_ecef variable and
+ * sets #base_pos_known to `true`.
+ */
+static void base_pos_callback(u16 sender_id, u8 len, u8 msg[], void* context)
+{
+  (void) context; (void) len; (void) msg; (void) sender_id;
+
+  double llh_degrees[3];
+  double llh[3];
+  memcpy(llh_degrees, msg, 3*sizeof(double));
+
+  llh[0] = llh_degrees[0] * D2R;
+  llh[1] = llh_degrees[1] * D2R;
+  llh[2] = llh_degrees[2];
+
+  chMtxLock(&base_pos_lock);
+  wgsllh2ecef(llh, base_pos_ecef);
+  base_pos_known = true;
+  chMtxUnlock();
+}
+
+/** SBP callback for the old style observation messages.
+ * Just prints a deprecation warning. */
+static void obs_old_callback(u16 sender_id, u8 len, u8 msg[], void* context)
+{
+  (void) context; (void) len; (void) msg; (void) sender_id;
+  printf("Receiving an old deprecated observation message.\n");
+  printf("Please update your base station firmware.\n");
+}
+
+/** Update the #base_obss state given a new set of obss.
+ * First sorts by PRN and computes the TDCP Doppler for the observation set. If
+ * #base_pos_known is false then a single point position solution is also
+ * calculated. Next the `has_pos`, `pos_ecef` and `sat_dists` fields are filled
+ * in. Finally the #base_obs_received semaphore is flagged to indicate that new
+ * observations are available.
+ *
+ * \note This function is stateful as it must store the previous observation
+ *       set for the TDCP Doppler.
+ */
+static void update_obss(obss_t *new_obss)
+{
+  /* Ensure observations sorted by PRN. */
+  qsort(new_obss->nm, new_obss->n,
+        sizeof(navigation_measurement_t), nav_meas_cmp);
+
+  /* Lock mutex before modifying base_obss.
+   * NOTE: We didn't need to lock it before reading in THIS context as this
+   * is the only thread that writes to base_obss. */
+  chMtxLock(&base_obs_lock);
+
+  /* Create a set of navigation measurements to store the previous
+   * observations. */
+  static u8 n_old = 0;
+  static navigation_measurement_t nm_old[MAX_CHANNELS];
+
+  /* Fill in the navigation measurements in base_obss, using TDCP method to
+   * calculate the Doppler shift. */
+  base_obss.n = tdcp_doppler(new_obss->n, new_obss->nm,
+                             n_old, nm_old, base_obss.nm);
+  /* Copy over the time. */
+  base_obss.t = new_obss->t;
+
+  /* Copy the current observations over to nm_old so we can difference
+   * against them next time around. */
+  memcpy(nm_old, new_obss->nm,
+         new_obss->n * sizeof(navigation_measurement_t));
+  n_old = new_obss->n;
+
+  /* Reset the `has_pos` flag. */
+  u8 has_pos_old = base_obss.has_pos;
+  base_obss.has_pos = 0;
+  /* Check if the base station has sent us its position explicitly and if so us
+   * that. No need to lock before reading here as base_pos_* is only written
+   * from this thread (SBP). */
+  if (base_pos_known) {
```

Comment:
where is the sent position coming from? If it moves around (e.g. single point position), it will kill the low latency mode.